pkg/events/derive/symbols_collision.go

```diff
@@ -0,0 +1,288 @@
+	processLoadedObjectsLRU, _ := simplelru.NewLRU(1024, lruCallback)
+	objectCollisionsLRU, _ := simplelru.NewLRU(1024, lruCallback)
+	symbolsBlackListMap := make(map[string]bool)
+	for _, sym := range symbolsBlackList {
+		symbolsBlackListMap[sym] = true
+	}
+	symbolsWhiteListMap := make(map[string]bool)
+	for _, sym := range symbolsWhiteList {
+		symbolsWhiteListMap[sym] = true
+	}
+	return &SymbolsCollisionArgsGenerator{
+		soLoader:          soLoader,
+		processesSOsCache: processesLoadedObjectsCache{processLoadedObjectsLRU},
+		collisionsCache:   sharedObjectsCollisionsCache{objectCollisionsLRU},
+		symbolsBlackList:  symbolsBlackListMap,
+		symbolsWhiteList:  symbolsWhiteListMap,
+	}
+}
+
+func (soColGen *SymbolsCollisionArgsGenerator) deriveArgs(event trace.Event) ([][]interface{}, []error) {
+	switch events.ID(event.EventID) {
+	case events.SharedObjectLoaded:
+		return soColGen.handleSOLoaded(event)
+	case events.SchedProcessExec:
+		return soColGen.handleExec(event)
+	default:
+		return nil, []error{fmt.Errorf("received unexpected event - \"%s\"", event.EventName)}
+	}
+}
+
+// handleSOLoaded check when a shared object is loaded into a process if some of its exported symbols collide with
+// previously loaded shared object.
+// An event will be created for each SO it has collisions with.
+// The main efficiency features it uses is caching examined SOs symbols and collision check results.
+// The hope is that after few processes are executed, all major libraries in the system will be cached already
+// and most libraries combinations collisions check results will be cached too.
+func (soColGen *SymbolsCollisionArgsGenerator) handleSOLoaded(event trace.Event) ([][]interface{}, []error) {
+	loadingObjectInfo, err := getSharedObjectInfo(event)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	processLoadedObjects, ok := soColGen.processesSOsCache.GetProcessLoadedObjects(event.HostProcessID)
+	if !ok {
+		processLoadedObjects = []sharedobjs.ObjInfo{}
+	}
+	newProcessLoadedObjects := append(processLoadedObjects, loadingObjectInfo)
+	soColGen.processesSOsCache.SetProcessLoadedObjects(event.HostProcessID, newProcessLoadedObjects)
+
+	// Exported symbols will be updated if needed in the findSOCollisions method
+	loadingObject := loadingSOInstance{ObjInfo: loadingObjectInfo}
+
+	var collisionEventsArgs [][]interface{}
+	var errs []error
+	for _, lsoInfo := range processLoadedObjects {
+		collisions, err := soColGen.findSOCollisions(&loadingObject, lsoInfo)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if len(collisions) > 0 {
+			collisionEventsArgs = append(collisionEventsArgs, []interface{}{loadingObjectInfo.Path, lsoInfo.Path, collisions})
+		}
+	}
+
+	return collisionEventsArgs, errs
+}
+
+// handleExec delete saved process loaded objects in case of execve, because entire memory is overwritten in this case
+func (soColGen *SymbolsCollisionArgsGenerator) handleExec(event trace.Event) ([][]interface{}, []error) {
```

Comment:
Do we really need the handle Exec?

You're caching the []sharedobjs.ObjInfo per PID number in the cache. Linux will have 4 Million PIDs, which is much higher than the size of the LRU.

I guess you're thinking if we're using filters then the LRU cache might still have an entry for that PID ?

Reply:
What I am thinking is that someone will execute a binary twice in a row using the same process.
This way the cache will maintain the first execution SOs, but even though they are not relevant because the process memory was overwritten.
This should reduce false-positives.